File: src/sources/bluesky.js
const SEARCH_TERMS = [
  "AI agent",
  "tool calling",
  "coding agent",
  "MCP server",
  "function calling LLM",
];

export async function fetchBluesky() {
  const since = new Date(Date.now() - 24 * 60 * 60 * 1000).toISOString();
  const results = [];

  for (const term of SEARCH_TERMS) {
    try {
      const params = new URLSearchParams({
        q: term,
        since,
        sort: "top",
        limit: "25",
      });
      const res = await fetch(
        `https://public.api.bsky.app/xrpc/app.bsky.feed.searchPosts?${params}`
      );
      if (!res.ok) continue;
      const data = await res.json();

      for (const post of data.posts || []) {
        const likes = post.likeCount || 0;
        const reposts = post.repostCount || 0;
        const engagement = likes + reposts;
        if (engagement < 5) continue;

        const handle = post.author?.handle || "unknown";
        const rkey = post.uri?.split("/").pop();
        const url = `https://bsky.app/profile/${handle}/post/${rkey}`;
        const text = post.record?.text || "";
        const title =
          text.length > 140 ? text.slice(0, 140) + "..." : text;

        results.push({ title, url, source: "Bluesky", engagement });
      }
    } catch {
      // skip failed searches
    }
  }

  // Deduplicate by URL and sort by engagement
  const seen = new Set();
  return results
    .filter((r) => {
      if (seen.has(r.url)) return false;
      seen.add(r.url);
      return true;
    })
    .sort((a, b) => b.engagement - a.engagement)
    .slice(0, 20);
}

File: ai_daily_digest/sources/bluesky.py
import os
import httpx
from datetime import datetime, timedelta, timezone
from atproto import Client
from dotenv import load_dotenv

load_dotenv()

SEARCH_TERMS = [
    "AI agent",
    "tool calling",
    "coding agent",
    "MCP server",
    "function calling LLM",
]

BSKY_API = "https://bsky.social/xrpc"

client = Client()
IDENTIFIER = os.getenv("BSKY_IDENTIFIER")
PASSWORD = os.getenv("BSKY_APP_PASSWORD")
client.login(IDENTIFIER, PASSWORD)


def fetch_bluesky():
    since = (datetime.now(timezone.utc) - timedelta(hours=24)).isoformat()
    results = []

    for term in SEARCH_TERMS:
        try:
            params = {"q": term, "since": since, "sort": "top", "limit": 25}
            res = client.app.bsky.feed.search_posts(params)

            for post in res.posts:
                likes = post.like_count
                reposts = post.repost_count
                engagement = likes + reposts
                if engagement < 5:
                    continue

                handle = post.author.handle
                rkey = post.uri.split("/")[-1]
                url = f"https://bsky.app/profile/{handle}/post/{rkey}"
                text = post.record.text
                title = text[:1000] + "..." if len(text) > 1000 else text

                results.append(
                    {"title": title, "url": url, "source": "Bluesky", "engagement": engagement}
                )
        except Exception as err:
            print(err)
            continue

    seen = set()
    deduped = []
    for r in results:
        if r["url"] not in seen:
            seen.add(r["url"])
            deduped.append(r)
    deduped.sort(key=lambda r: r["engagement"], reverse=True)
    return deduped[:20]

if __name__ == "__main__":
    fetch_bluesky()
